devito/core/gpu_openmp.py

```diff
@@ -211,6 +221,83 @@ def _(iet):
 
         return iet, {}
 
+@iet_pass
+def initialize(iet, **kwargs):
+    """
+    Initialize the OpenMP environment.
+    """
+
+    @singledispatch
+    def _initialize(iet):
+        comm = None
+        devicenum = Symbol(name='devicenum')
+        for i in iet.parameters:
+            if isinstance(i, MPICommObject):
+                comm = i
+                break
+            
+        if comm is not None:
+            rank = Symbol(name='rank')
+            rank_decl = LocalExpression(DummyEq(rank, 0))
+            rank_init = Call('MPI_Comm_rank', [comm, Byref(rank)])
+
+            ngpus = Symbol(name='ngpus')
+            call = Function('omp_get_num_devices')()
+            ngpus_init = LocalExpression(DummyEq(ngpus, call))
+            
+            devicenum_init = LocalExpression(DummyEq(devicenum, rank % ngpus))
+
+            body = [rank_decl, rank_init, ngpus_init, devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0+MPI setup'),
+                    body=body,
+                    footer=(c.Comment('End of OpenMP5.0+MPI setup'), c.Line()))
+        else:
+            devicenum_init = LocalExpression(DummyEq(devicenum, 0))
+            body = [devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0 setup'),
```

Comment:
duplicate code?

Reply:
I don't think so. The code in _if_ statement is for MPI+GPU offloading, and the code in _else_ statement is for GPU offloading without MPI. They have different header, body, and footer.